### PR TITLE
generate a patch for each directory

### LINKED ISSUE #1650: generating  upstream patch script needs improvement
### Context
When using `scripts/generate-upstream-patch.sh` to generate patch for pallets, currently it generates patch for the whole pallet repo. As our local pallets and primitives have different folder structure, compared to the upstream pallets, the patch couldn't be applied smoothly. 

Reference #1607 , there was pallets update from `50cac49` to `75e57c5` of upstream repo. Create a patch based on the difference of `50cac49` to `75e57c5` of upstream pallet repo, then apply it. Patch goes to wrong places.

### Task
Find a solution to apply the patch. e.g.: create different patch for different folders (pallets.patch, primitives.patch, etc.)

### Acceptance criteria
What's the definition of done, if applicable?

---
:heavy_check_mark: Please set appropriate **labels** and **assignees** if applicable.

### PR DESCRIPTION
**Pull request by outofxxx**
resolved #1650 

* generate a patch for each pallets repo directory

About `allow-empty` option:
!!! Need to upgrade to the latest git version to use this option.

```shell
$ sudo add-apt-repository ppa:git-core/ppa
$ sudo apt-get update
$ sudo apt-get install git
```

### CODE REVIEW

File: scripts/generate-upstream-patch.sh [hunk 1]
```diff
@@ -56,116 +51,148 @@ generate_upstream_patch() {
...
-	then
-		echo "  git am -3 --exclude=tee-worker/Cargo.lock --exclude=tee-worker/enclave-runtime/Cargo.lock --directory=tee-worker < tee-worker/upstream.patch"
-	fi
+    echo "  pallets   > git apply -p1 -3 --directory=pallets $ROOTDIR/pallets/pallets_xxx.patch"
+	echo "  primitives> git apply -p1 -3 $ROOTDIR/pallets/primitives_xxx.patch"
+
+    echo ""
+	echo "after that, please:"
+	echo "- pay special attention: "
+    echo "  * ALL changes/conflicts from pallets/upstream.patch should ONLY apply into:"
```
Comment (BillyWooo) [inline scripts/generate-upstream-patch.sh]: `pallets/upstream.patch` It's not valid anymore.

File: scripts/generate-upstream-patch.sh [hunk 2]
```diff
@@ -2,46 +2,41 @@
...
+   echo "cleaned up $1"
 }
 
-check_upstream() {
-	local TARGET=$1
-	local UPSTREAM_URL="$UPSTREAM_URL_PREFIX/$TARGET"
-	if [ "$(git remote get-url upstream_$TARGET 2>/dev/null)" != "$UPSTREAM_URL" ]; then
-		git remote add upstream_$TARGET $UPSTREAM_URL
-	fi
+function usage() {
```
Comment (BillyWooo) [inline scripts/generate-upstream-patch.sh:13]: It's not possible to generate patch for all at one time?
Normally `[ ]` represents optional parameters, `< >` represents madentary parameters.

**Review by BillyWooo — CHANGES_REQUESTED** [review]
Please take care about the indent. Either space or tabs. Recommend `tabs`.

File: scripts/generate-upstream-patch.sh [hunk 3]
```diff
@@ -56,116 +51,148 @@ generate_upstream_patch() {
...
+then
+	generate_pallets_patch "$@"
+    apply_pallets_tips
+fi
+
+if [[ $HAS_WORKER == "true" ]]
+then
+    generate_worker_patch "$@"
+    apply_woker_tips
+fi
```
Comment (Kailai-Wang) [inline scripts/generate-upstream-patch.sh]: From this logic it seems that `HAS_PALLETS` and `HAS_WORKER` can't be true simultaneously.
Why not (if I'm not mistaken):
```
case "$1" in
    p)
        generate_pallets_patch "$@"
        apply_pallets_tips
        ;;
    w)
        generate_pallets_patch "$@"
        apply_woker_tips
        ;;
    *)
        usage; exit 1 ;;
esac

File: scripts/generate-upstream-patch.sh [hunk 4]
```diff
@@ -2,46 +2,41 @@
...
 #
-# A patch will be generated under ./<TARGET_DIR>/upstream.patch
-generate_upstream_patch() {
-	local TARGET=$1
+# Patches will be generated under ./<TARGET_DIR>/
+function generate_worker_patch() {
+	local TARGET="worker"
 	local TARGET_DIR="$ROOTDIR/$TARGET"
+    echo "TARGET: $1"
+    echo "TARGET_DIR: $TARGET_DIR"
```
Comment (arieljbirnbaum) [inline scripts/generate-upstream-patch.sh]: :nail_care: the indentation is not consistent; are you planning to remove the `echo` statements later?

File: scripts/generate-upstream-patch.sh [hunk 5]
```diff
@@ -56,116 +51,148 @@ generate_upstream_patch() {
...
-	if [ "$HAS_PALLETS" == "true" ]
-	then
-		echo "  git am -3 --directory=pallets < pallets/upstream.patch"
-	fi
-	if [ "$HAS_WORKER" == "true" ]
-	then
-		echo "  git am -3 --exclude=tee-worker/Cargo.lock --exclude=tee-worker/enclave-runtime/Cargo.lock --directory=tee-worker < tee-worker/upstream.patch"
-	fi
+    echo "  pallets   > git apply -p1 -3 --directory=pallets $ROOTDIR/pallets/pallets_xxx.patch"
+	echo "  primitives> git apply -p1 -3 $ROOTDIR/pallets/primitives_xxx.patch"
```
Comment (arieljbirnbaum) [inline scripts/generate-upstream-patch.sh]:
```suggestion
     echo " # Pallet patches:"
     echo " git apply -p1 -3 --directory=pallets $ROOTDIR/pallets/pallets_xxx.patch"
     echo " # Primitive patches:"
	echo " git apply -p1 -3 $ROOTDIR/pallets/primitives_xxx.patch"
```
:nail_care:

File: scripts/generate-upstream-patch.sh (hunk 2 above)
Comment (outofxxx) [reply scripts/generate-upstream-patch.sh:13]: i think it's better to generate seperately.

**Comment by Kailai-Wang** [comment]
What's the status of this PR? Is it read to be reviewed again?

**Review by BillyWooo — APPROVED** [review]
Be careful about the git version. Current apt installed version on ubuntu doesn't support `--allow-empty` flag.

**Pull request merged**